Add introduction to the user manual

sskeirik opened the pull request:
This is a first attempt at writing an introduction to the user manual that we now have.

Note that I am using this issue (#1585) as an overall reference for what to write and in what order.

Commit:
add introduction to user manual

sskeirik committed:
Merge branch 'master' into manual_intro

Commit:
fix typo in user manual

nishantjr commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,81 @@
+3.  Analyzing K Specs (e.g. theorem proving, bounded model checking)
+
+Different tools in the K Framework have different levels of maturity.
+The most well-supported and broadly used tools are the following:
```
```suggestion
The main *user-facing* K tools include:
```

nishantjr commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,81 @@
+Different tools in the K Framework have different levels of maturity.
+The most well-supported and broadly used tools are the following:
+
+-   `komile` - the K compiler driver
```
```suggestion
-   `kompile` - the K compiler driver
```

nishantjr commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,81 @@
+to peruse the
+[growing literature on K and matching logic](https://fsl.cs.illinois.edu/projects/pl/index.html).
+We will consider the manual complete when it provides a complete description of
+all well-supported K tools and features.
```
```suggestion
This user manual is designed to be a tool reference. In particular, it is not
desgined to be a tutorial on how to write K specifications or to teach the
logical foundations of K. New K users should consult our dedicated
[K tutorial](https://kframework.org/k-distribution/k-tutorial/),
or the more language-design oriented [PL tutorial](https://kframework.org/k-distribution/pl-tutorial/).
Researchers seeking to learn more about the logic underlying K are encouraged
to peruse the
[growing literature on K and matching logic](https://fsl.cs.illinois.edu/projects/pl/index.html).
We will consider the manual complete when it provides a complete description of
all well-supported K tools and features.
```

sskeirik committed:
Apply suggestions from code review

Co-authored-by: Nishant Rodrigues <nishantjr@gmail.com>

Commit:
cleanup objectives description in intro

Commit:
add more detailed K Framework overview

sskeirik committed:
Merge branch 'master' into manual_intro

Commit:
reorganize intro to put high-level description first

Commit:
fix the K parsing/syntax description in the intro

Commit:
add a transition sentence

Commit:
improve legend on k process diagram

Commit:
improve description of K syntax/parsing

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+specifications.*
+
+Typically, programming language tools are developed as follows: for each
+language *L*, a separate team develops tool *T* (e.g. a compiler, interpreter,
```
I don't know that we want to use this type of capital letter style here for something like this. It's better just to use plain English

sskeirik replied:
I rewrote this section to use plain english.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+parser, symbolic execution engine, etc). Thus, for *N* different languages and
+*M* different tools, we have *N x M* different teams and implementations.
+
+K approaches the problem differently, by imagining how each of the *M*
```
same here

sskeirik replied:
I rewrote this section to use plain english.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+The main *user-facing* K tools include:
+
+-   `kompile` - the K compiler driver
+-   `kast` - the stanadlone K parser and abstract syntax tree (AST)
```
let's replace this with `kparse`

sskeirik replied:
Done.

dwightguth commented on USER_MANUAL.md:
````diff
@@ -1,12 +1,355 @@
+*import* other modules, giving rise to a hierarchy of files and modules. We
+give an intuitive sketch of the two levels of grouping in the diagram below:
+
+```k
````
this is not k code so the `k` label is not appropriate on this code fence

sskeirik replied:
Removed.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+Here we provide an overview of the top-level K syntax using a K-style BNF
+notation. Note that this is not a full definition; we focus instead on the
+overall structure of the specification. Syntactic categories which are not
+defined are denoted by names surrounded by curly braces (`{}`). In particular,
```
if we're going to be giving a grammar, we shouldn't make it incomplete like this. It's possible to define the syntax of all of these identifiers and such fairly readily.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+
+1.  In the first stage, system primitive declarations are gathered and a parser
+    is constructed.
+2.  In the second stage, the system primitive parser is used to parse
```
better to talk about outer and inner parsing as that is a much more endemic nomenclature to the project.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+3.  Finally, a third stage parser is constructed containing both primitive and
+    configuration syntax is constructed which is used to parse contexts,
+    context aliases, and rules. In our BNF grammar below, we use the words
+    `{Bubble}` and `{BubbleWithRewrites}` to describe parsers which contain the
```
this is incorrect; there is only one type of bubble. What may appear inside that bubble is not part of the outer syntax of K; it's part of a separate grammar that is generated from the user-defined syntax.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  Module       ::= "module" {ModuleId} AttributeSet ImportList SentenceList "endmodule"
+
+  AttributeSet ::= "" | "[" Attributes "]"
+  Attributes   ::= Attributes "," {Attribute}
```
this is not correct as your grammar here does not allow for the possibility of a sentence with only a single attribute.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+
+  GenericNotation    ::= StringOrSortIdList
+  StringOrSortIdList ::= StringOrSortId | StringOrSortId StringOrSortIdList
+  StringOrSortIdList ::= {String} | {SortId}
```
this is also not correct as you are not allowing for regular expression terminals

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  NeIdList                 ::= {Id} | {Id} NeIdList
+  Associativity            ::= "left" | "right" | "non-assoc"
+
+  PriorityDeclaration ::= "syntax" "priorities" PriorityList
```
this can also be written as "syntax priority"

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  Associativity            ::= "left" | "right" | "non-assoc"
+
+  PriorityDeclaration ::= "syntax" "priorities" PriorityList
+  PriorityList        ::= NeIdList ">" PriorityList
```
this is also not correct as your grammar does not allow a sentence with only two priority levels

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  PriorityList        ::= NeIdList ">" PriorityList
+                        | NeIdList ">" NeIdList
+
+  ConfigDeclaration ::= "configuration" NeCellDeclList
```
this is not really how configuration parsing works.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  RuleDeclaration   ::= "rule" OptName {BubbleWithRewrites} RequiresCondition EnsuresCondition AttributeSet
+  OptName           ::= "" | Name
+  Name              ::= "[" {Id} "]" ":"
+  RequiresCondition ::= "" | "requires" {Bubble}
```
this is also not how bubbles in rules work

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,355 @@
+  EnsuresCondition  ::= "" | "ensures" {Bubble}
+
+  ContextDeclaration      ::= "context" {BubbleWithRewrites} RequiresCondition AttributeSet
+  ContextAliasDeclaration ::= "context" "alias" Name {Bubble} AttributeSet
```
context aliases can have requires clauses.

sskeirik replied:
I removed this section from this PR because it needs more polish.

dwightguth reviewed (changes requested):
I have some small low-level concerns with wording, but the biggest problem is that your grammar is simply not a good description of how outer parsing works, even if we set aside the individual minor bugs in how it's defined.

sskeirik commented:
@dwightguth Thanks for your review comments. Re: bubble parsing: I don't fully understand how this works, as you saw, what I wrote was my best attempt. Two obvious possible paths forward on this:

1.  throw out the grammar section and keep the other stuff
2.  try and fix the grammar section by fixing the bubble description

If we want to pursue option (2), I would appreciate some help on:

1.  pointers on where I can find the syntax of valid K identifiers for sorts, cells, record selectors, modules, etc.
2.  help understanding how bubble parsing works

dwightguth commented:
Generally speaking all the syntax for the outer parsing can be found in `kernel/src/main/javacc/Outer.jj` and all the syntax for inner parsing can be found in `k-distribution/include/kframework/builtin/kast.k` and `kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java`. It shouldn't be too hard to find the syntaxes for the identifiers since these should all be inside Outer.jj or kast.k in definitions of formal grammars.

Bubble parsing is more complex. Right now, here is (roughly) how it works:

A bubble starts immediately following the `rule`, `configuration`, `context`, `context alias`, or `claim` keywords and continues until the next occurrence of the `syntax`, `endmodule`, `rule`, `claim`, `configuration`, or `context` keywords. A bubble can contain any characters; parsing of that content happens later.

The first pass of bubble parsing takes place inside the outer parser itself; we look for a rule label in the form of the `"(?s)\\s*\\[\\s*([^\\[\\]\\_\\ \\n\\r\\t]+)\\s*\\]\\s*:\\s*(.*)"` Java regular expression, and extract it out from the bubble and put it in the attributes of the sentence as the `label` attribute. We also try to find a spot in the sentence starting with a `[` at which it is possible to derive a set of attributes immediately followed by the end of the bubble. If we find one, that becomes the attributes for that sentence, and everything left over after this first pass of bubble parsing becomes the final bubble, ready for inner parsing. If we don't find one, the bubble extends to the very end and the sentence has no explicit attributes.

Once this pass of bubble parsing is complete, we have the attributes of the sentence, the type of the sentence, and a string containing the body of the sentence itself. We then generate a scanner and parser in order to parse that bubble, and everything inside gets parsed as a term of sort `#RuleContent`. The grammar used to generate this scanner and parser is augmented with the syntax of configurations (as defined by the `CONFIG-CELLS` module in the case of configurations and the `RULE-CELLS` modules in the case of contexts, context aliases, rules, and claims) and the syntax of K (as defined by the `K` module). We also use the `#Layout` sort to decide what characters the scanner should treat as whitespace or comments. If the `#Layout` sort is not defined by the user, we use the `DEFAULT-LAYOUT` module.

One final component gets mixed into the grammar generation, and that is the code mentioned above in `RuleGrammarGenerator.java`. This includes various pieces of logic that munge the grammar or add certain productions that cannot be easily expressed in K, like casts or sort predicates. RuleGrammarGenerator also takes care of any necessary internal munging in order to make the grammar into a form that our implementation can handle; you shouldn't need to worry about those details when describing how parsing works though.

dwightguth commented:
All that being said, I have no objection if you want to remove that part of the manual and handle it later.

Commit:
remove syntax/parsing description from manual

Commit:
address review comments

sskeirik commented:
@dwightguth I believe I have addressed your review comments on wording and deleted the syntax section because it needs more polish. I would appreciate a re-review. Thanks.

sskeirik committed:
Merge branch 'master' into manual_intro

sskeirik committed:
Merge branch 'master' into manual_intro

nishantjr commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,242 @@
+input and produces a tool as output. The end result is that the exercise of
+designing new languages and their associated tooling is now reduced to
+developing a single language specfication from which we derive our tooling *for
+free*.
```
It's a little confusing that the tools are both "derived generically from a common specification" and that they take the "language spec as input".

```suggestion
K approaches the problem differently -- it generates each of these tools from a single language specification.
The work of programming language design and tool implementation are made separate concerns.
The end result is that the exercise of
designing new languages and their associated tooling is now reduced to
developing a single language specification from which we derive our tooling *for
free*.
```

nishantjr commented on USER_MANUAL.md:
````diff
@@ -1,12 +1,242 @@
+can be understood as a blackbox with the following inputs and outputs:
+
+```
+ K Compilation Process
++============================================================+
+|                     +---------+                            |
+|  K Specification ---| kompile |--> Kore Specification --+  |
+|                     +---------+                         |  |
++=========================================================|==+
+                                                          |
+ K Execution Process                                      |
++=========================================================|==+
+|                                                         |  |
+|             +-------------------------------------------+  |
+|             |                                              |
+|             |       +---------+                            |
+|  K Term ----+-------| kparse  |--> K Term                  |
+|             |       +---------+                            |
+|             |                                              |
+|             |       +---------+                            |
+|  K Term ----+-------|  krun   |--> K Term                  |
+|             |       +---------+                            |
+|             |                                              |
+|             |       +---------+                            |
+|  K Claims --+-------| kprove  |--> K Claims                |
+|                     +---------+                            |
+|                                                            |
++============================================================+
````
Kore has not been introduced yet, also `kparse` and `krun` take Programs as input.

```suggestion
 K Compilation Process
+============================================================+
|                     +---------+                            |
|  K Specification ---| kompile |--> Formal Specification +  |
|                     +---------+                         |  |
+=========================================================|==+
                                                          |
 K Execution Process                                      |
+=========================================================|==+
|                                                         |  |
|             +-------------------------------------------+  |
|             |                                              |
|             |       +---------+                            |
|  Program ---+-------| kparse  |--> K Term                  |
|             |       +---------+                            |
|             |                                              |
|             |       +---------+                            |
|  Program ---+-------|  krun   |--> K Term                  |
|             |       +---------+                            |
|             |                                              |
|             |       +---------+                            |
|  K Claims --+-------| kprove  |--> K Claims                |
|                     +---------+                            |
|                                                            |
+============================================================+
```

nishantjr commented on USER_MANUAL.md:
```diff
@@ -1,12 +1,242 @@
+where:
+
+-   process outlines are denoted by boxes with double-lined borders
+-   programs are denoted by boxes with single-lined borders
```
I think we can avoid confusion between programs in the programming language, and executables.

```suggestion
-   executables are denoted by boxes with single-lined borders
```

sskeirik committed:
Apply suggestions from code review

Co-authored-by: Nishant Rodrigues <nishantjr@gmail.com>

Commit:
add more explanatory information to the process diagram

nishantjr reviewed (approved):
Looks good.
